firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SessionGenerator.kt

```diff
@@ -36,35 +36,25 @@ internal data class SessionDetails(
  *
  * @hide
  */
-internal class SessionGenerator(private var collectEvents: Boolean) {
-  private var firstSessionId = ""
-  private var sessionIndex: Int = -1
+internal class SessionGenerator(private val collectEvents: Boolean) {
+  private val firstSessionId = generateSessionId()
+  private var sessionIndex = 0
 
-  private var thisSession: SessionDetails =
-    SessionDetails(
-      sessionId = "",
-      firstSessionId = "",
-      collectEvents,
-      sessionIndex,
-    )
+  var currentSession =
+    SessionDetails(sessionId = "", firstSessionId = "", collectEvents, sessionIndex = -1)
+    private set
 
-  // Generates a new Session ID. If there was already a generated Session ID
-  // from the last session during the app's lifecycle, it will also set the last Session ID
+  /** Generates a new session. The first session's sessionId will match firstSessionId. */
   fun generateNewSession(): SessionDetails {
-    val newSessionId = UUID.randomUUID().toString().replace("-", "").lowercase()
-
-    // If firstSessionId is set, use it. Otherwise set it to the
-    // first generated Session ID
-    firstSessionId = firstSessionId.ifEmpty { newSessionId }
-
-    sessionIndex += 1
-
-    thisSession =
-      SessionDetails(sessionId = newSessionId, firstSessionId, collectEvents, sessionIndex)
-
-    return thisSession
+    currentSession =
+      SessionDetails(
+        sessionId = if (sessionIndex == 0) firstSessionId else generateSessionId(),
+        firstSessionId,
+        collectEvents,
+        sessionIndex++,
+      )
+    return currentSession
   }
 
-  val currentSession: SessionDetails
-    get() = thisSession
+  private fun generateSessionId() = UUID.randomUUID().toString().replace("-", "").lowercase()
```

Comment:
It might not be worth it, but if we made this an interface like `SessionIdGenerator` and passed in, we could make the unit tests deterministic. Then have a seperate test to validate the generated session ids. What do you think?

Reply:
I like that idea. 2 suggestions:
 - You can pass in a function to the constructor instead of making a SessionIdGenerator class with a generate() method. Only reason is a function might be lighter weight.
  - This class / function that we pass in should only return the `UUID.randomUUID()` part of this statement. This `SessionGenerator` should be responsible for toString and `.replace`ing it so we can assert that the dashes are removed in tests.

Reply:
A separate interface just for sessionId sounds like an overkill. Do you think this structure is not testable? I think with the SessionDetails structure and generateNewSession, we should be able to cover all scenarios for testing.

Reply:
I was concerned about the randomness in the tests, and asserts on individual fields of the SessionDetails data class to avoid a big assert on the entire object since it contained randomness.

I played around with this. I made it take a function, not an interface, in the ctor. I left one test with the random UUIDs to verify it generates valid session ids.